[ENT-343] Add Terms of Service modal

replaceafill opened the pull request:
This PR allows the user to view the Terms of Service (TOS) without leaving the Register page. The TOS are now displayed in a modal that can be closed afterwards. The modal includes an `<iframe>` element that opens the `href` of the TOS link. The styles of the inserted page are modified to hide the navigational and footer areas.

**JIRA tickets**: [ENT-343](https://openedx.atlassian.net/browse/ENT-343)

**Dependencies**: None.

**Screenshots**:
![tos-modal](https://cloud.githubusercontent.com/assets/560781/25751478/5d13c16e-3172-11e7-87f2-63f88ac606e8.png)

**Sandbox URL**:

LMS: https://pr15038.sandbox.opencraft.hosting/
Studio: https://studio-pr15038.sandbox.opencraft.hosting/

I added some extra pagraphs to the sandbox `lms/templates/static_templates/honor.html` template so the modal can be scrollable.

**Merge deadline**: Friday May 5th, 2017.

**Testing instructions**:

1. Set up a devstack with this branch.
2. Go to the **Register** page.
3. Click the **Review the Terms of Service and Honor Code** (TOS) link.
4. The Terms of Service page should be displayed in a modal screen.
5. Clicking the close button or outside the content area should close the modal.

**Reviewers**
- [x] @itsjeyd 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @replaceafill! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=15038"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

itsjeyd commented on lms/static/js/student_account/register.js:
```diff
@@ -0,0 +1,111 @@
+            }),
+            content = $('<div>', {
+                'class': content_class,
+                role: 'dialog',
```
@replaceafill Missing quotes around `role`?

replaceafill replied:
I used quotes around `class` because it's a keyword. `role` is not so it can be used safely. The `aria-*` ones needs them because the dash is also an operator. I added the quotes anyway for consistency :)

itsjeyd replied:
@replaceafill Ah, right. Thanks for adding the quotes though, consistency is nice :)

itsjeyd commented on lms/static/js/student_account/register.js:
```diff
@@ -0,0 +1,111 @@
+    var hideModal = function(modal_selector) {
+        $(modal_selector).hide();
+        $('body').removeClass('open-modal');
+        $(modal_selector).find('iframe').remove();
```
@replaceafill Why is it necessary to remove the `iframe` completely? (genuine question)

replaceafill replied:
I was having trouble adjusting the height of the embedded page when the dialog was closed and reopen. This was the quick/dirty solution I used.

itsjeyd commented on lms/static/js/student_account/register.js:
```diff
@@ -0,0 +1,111 @@
+            if ((event.keyCode === 27) && ($(modal_selector).is(':visible'))) {
+                // hide the modal if ESC is pressed
+                hideModal(modal_selector);
+                event.preventDefault();
```
@replaceafill Any specific reason for calling `preventDefault` last (instead of calling it before any other logic)?

itsjeyd commented on lms/static/js/student_account/register.js:
```diff
@@ -0,0 +1,111 @@
+            content_selector = '.' + content_class,
+            modal_class = 'modal',
+            modal_selector = '.' + modal_class;
+        $('body').on('click', tos_link_selector, function(event) {
```
@replaceafill Please add some vertical whitespace to this part of the code as well.

bradenmacdonald commented:
Is there some way to do this without an iframe?

Also, I'm pretty sure there are not supposed to be dotted/dashed lines. Those look like part of the mockup, not part of the desired end UI.

replaceafill commented:
@bradenmacdonald I went with the iframe because the TOS URL can be customized through the `MKTG_URLS` env variable. So I wasn't sure how much control we would have over the templates.

replaceafill commented on lms/static/js/spec/student_account/register_spec.js:
```diff
@@ -366,6 +369,43 @@
+                it('displays a modal with the terms of service', function() {
+                    createRegisterView(this);
+
+                // Check there is no modal container initially
```
Yeah. It's an existing convention in that spec file (it got my attention too). See the [comments in the test for registering a user]( https://github.com/edx/edx-platform/blob/master/lms/static/js/spec/student_account/register_spec.js#L250-L271).

replaceafill commented:
@itsjeyd I finished addressing your observations and left a few comments too. Let me know if I missed anything.

I also updated the screenshot in the description.

bradenmacdonald commented:
@replaceafill I think the close link should be blue, as shown in the mockup?

replaceafill commented:
@bradenmacdonald Fixed.

itsjeyd commented on lms/static/js/student_account/tos_modal.js:
```diff
@@ -0,0 +1,119 @@
+            src: link.href,
+            load: function() {
+                var iframeHead = $(this).contents().find('head');
+                // Overwrite styles in child page hiding top navigation and footer
```
@replaceafill Nit: This sounds like the styles in the child page are hiding the top navigation and footer, and that that's the behavior we're trying to override. How about changing the wording as follows:

> Overwrite styles in child page to hide top navigation and footer

replaceafill commented:
@brittneyexline I [disabled tab index cycling](https://github.com/replaceafill/edx-platform/blob/replaceafill/ENT-343/lms/static/js/student_account/tos_modal.js#L69-L98) of the background form when the modal is displayed and updated the sandbox. Could you please check again? Thanks!

douglashall commented:
It looks like the modal header isn't quite right when the header title is long:

<img width="395" alt="screen shot 2017-05-09 at 2 56 50 pm" src="https://cloud.githubusercontent.com/assets/224634/25867678/537accb6-34c8-11e7-8e32-1455e8ae0198.png">

Is there anything we can do about that?

douglashall commented on lms/djangoapps/static_template_view/views.py:
```diff
@@ -46,7 +46,11 @@ def render(request, template):
+        context = {}
+        # This is necessary for the dialog presented with the TOS in /register
+        if template == 'honor.html':
+            context['allow_iframing'] = True
```
What does setting this context variable do?

replaceafill replied:
@douglashall It allows the /honor view to be displayed inside the modal iframe. Otherwise [this part of the main.html template](https://github.com/edx/edx-platform/blob/master/lms/templates/main.html#L44-L53) redirects to the full /honor URL.

douglashall commented on lms/templates/student_account/login_and_register.html:
```diff
@@ -22,6 +22,7 @@
             newrelic.addPageAction('xfinished');
         }
     </%static:require_module>
+    <script type="text/javascript" src="${static.url('js/student_account/tos_modal.js')}"></script>
```
We should make the inclusion of this Javascript file configurable by site so that we can apply this new functionality on a per-site basis.

You can do this using a `SiteConfiguration` value, like so:
```
<%!
from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
%>
% if configuration_helpers.get_value('DISPLAY_TOS_IN_MODAL_ON_REGISTRATION_PAGE', False):
<script type="text/javascript" src="${static.url('js/student_account/tos_modal.js')}"></script>
% endif
```

replaceafill replied:
@douglashall If this is a site configuration, where would the `DISPLAY_TOS_IN_MODAL_ON_REGISTRATION_PAGE` setting be defined? I tried on my devstack's `lms.envs.private` module but I think that doesn't work.

douglashall replied:
`SiteConfiguration` is a database model. You would set the value using [Django admin](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/sites/configure_site.html).

replaceafill replied:
Thanks @douglashall. I implemented it using your suggestion. One thing I noticed though is that for the site configuration setting to take effect I need to restart my devstack/sandbox. Is that normal or maybe I haven't set my site correctly?

douglashall replied:
Awesome, thanks! SiteConfiguration is cached in memory for 5 minutes. So you can either restart or wait the 5 minutes.

chris-mike commented:
I have a few suggestions:
1. Re: title of page/modal: Please see the mock with regards to the title of the TOS modal and the page content. The title should be "edX Terms of Service and Honor Code." It should wrap. It should always be visible.
2. If there is a "title" w/in the iframe (or whatever you use to contain the body), it should not be bigger than the actual title. In the example you've put together, "Honor Code" does not need to be there at all. The title is "edX Terms of Service...."
3. Why do we have the black background behind the modal? I'd prefer we not have something so dark, but I'm not sure where this is coming from. A central style sheet?
4. Have the modal consume the page when it is in narrowest view.  Neither the black background nor the extra padding is necessary here.
5. Can we use standard font treatment here? The line height seems too tight. Also, the headlines use a different font. Can we use the font treatment seen here: https://www.edx.org/edx-terms-service
6. Focus should stay w/in the modal. It shouldn't go to the browser's URL field. 

Also, checking in with @cptvitamin regarding scrolling. When focus is on "close" should I be able to scroll?  If there weren't a link in the content body, i couldn't get to that part of the content with keyboard only.

![tos mobile](https://cloud.githubusercontent.com/assets/16578372/25870286/92053648-34d1-11e7-9fdf-4e526eb62ccb.png)

replaceafill commented:
@chris-mike @douglashall These are some changes I just made:

1. Changed the title of the modal to "edX Terms of Service and Honor Code" and it should not collapse with the close button anymore.
2. Used a lighter background color. I was using [black with 75% opacity](https://github.com/edx/edx-platform/blob/master/lms/static/sass/partials/base/_variables.scss#L46-L50). Now it uses 25% opacity.
3. Removed the padding on the sides when the dialog gets very narrow.

@chris-mike Regarding some of your suggestions:

2 and 5. The modal is just inserting the actual tos/honor view using an iframe. Since this view can be customized though the MKTG_URLS env variable, the modal has no much control on the content or the formatting of that view. "Honor Code" is displayed because it's in the content area of the full view (https://pr15038.sandbox.opencraft.hosting/honor). All I've done in the dialog is to hide the navigation area and the footer.
6 and 7. I think we could trap the [TAB] key to prevent cycling through the address bar and the UP/DOWN arrow keys to make the iframe to scroll accordingly. I'll wait on @cptvitamin feedback to implement this.

Thanks a lot for your observations.

cptvitamin commented:
The only issue I had is that once focus was inside of the iframe, I could not close the dialog with the ESC key. Once I got to the end of the TOS, I had to SHIFT+TAB all the way back up to the close button to dismiss the dialog.  Is there a way we can listen for the ESC key even while focus is inside the iframe?

bradenmacdonald commented:
@replaceafill Can we test/preview how this will look on edx.org? There the TOS is an external URL, https://www.edx.org/edx-terms-service

replaceafill commented:
@bradenmacdonald Here's a screenshot with a local copy of the TOS on edx.org.

![edx-tos](https://cloud.githubusercontent.com/assets/560781/25981156/ac1d69f4-3690-11e7-8d90-d7e4ac3eb0bf.png)

The dialog behaves the same as with the default honor.html template.

replaceafill commented:
@cptvitamin I added a listener to the iframe's body so now pressing ESC works when either the modal or the iframe have the focus.

cptvitamin commented:
👍

chris-mike commented:
Thanks for the changes. 
I would prefer that we have more control over the display of titles and headers. Currently on the edx.org TOS, we use an H1 for the introdoctory line. The title of the page should be "edX TOS & Honor Code). Then maybe some special visual treatment for "On October..."  Not sure how we can accomplish this.

Could you please do the following:
1. B/c we get to have title on top of a "title", can you add a little extra padding at the bottom of the header (for visual separation), and middle align the "edx TOS and HC" text.?
2. It's not clear how tall the "close" click target is. All click targets should be at least 48px tall. If you add extra padding (per #1), then this might be about the same height as the header frame.  
3. Can you make the modal taller by about 20%.

replaceafill commented:
Thanks for your detailed instructions @chris-mike. I've addressed them all. I also set up a copy of the edX's Terms of Service page in the sandbox. Unfortunately because of the hackish way I'm doing that, the sandbox navigation breaks, but you can access the Register page directly to see the modal changes:

https://pr15038.sandbox.opencraft.hosting/register

You can also see the copy of the TOS being used here:

https://pr15038.sandbox.opencraft.hosting/edx-terms-service.html

Let me know what you think.

douglashall commented:
@replaceafill @haikuginger @bradenmacdonald This one seems like it is ready for a rebase/squash/merge?

replaceafill commented:
@douglashall @haikuginger @bradenmacdonald I just rebased/squashed my changes. Could we run jenkins of this?

douglashall commented:
jenkins run all

douglashall commented:
jenkins run all

estute commented:
jenkins run all

replaceafill commented:
@estute @douglashall I apologize for not catching the run_quality issue before requesting the jenkins run. This should pass now.

douglashall commented:
jenkins run all

haikuginger commented:
Okay, we've got good CI and green CI - @douglashall, are we set to merge? @itsjeyd, since you're the OpenCraft reviewer on this, go ahead and merge if @douglashall says so.

douglashall commented:
@haikuginger We need to get the tests to run again before merging. The green is indicating the successful reviews, but no tests have run.

@estute Can you help us run the tests on this again?

estute commented:
jenkins run all

douglashall commented on lms/templates/student_account/login_and_register.html:
```diff
@@ -1,6 +1,7 @@
 <%!
```
You will need to add a page directive which escapes all strings by default to the top of this template in order to resolve the [safe template linter](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#safe-template-linter) quality [errors](https://build.testeng.edx.org/job/edx-platform-quality-pr/34621/artifact/edx-platform/reports/safelint/safelint.report) you are getting on this PR.
```
<%page expression_filter="h"/>
```
For more details, see [this](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#set-html-escaping-filter-as-default).

replaceafill replied:
@douglashall Thanks! Do you have any idea why my devstack doesn't catch these issues when I run `paver run_quality` (here's the full output: https://pastebin.com/QTsv5yFy)? Maybe there's some caching going on? I tried running `paver update_assets lms --settings=devstack && paver update_assets cms --settings=devstack` first.

douglashall replied:
I think this quality check is not included in `run_quality` for some reason. You have to explicitly run `paver run_safecommit_report`. See [safe template linter](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#safe-template-linter) for more details.

replaceafill replied:
I realized `paver run_quality` just generates the reports. I needed to run `paver run_eslint` before. Hopefully it'll pass this time.

douglashall commented on lms/templates/student_account/login_and_register.html:
```diff
@@ -22,6 +23,9 @@
     </%static:require_module>
+    % if configuration_helpers.get_value('DISPLAY_TOS_IN_MODAL_ON_REGISTRATION_PAGE', False):
+    <script type="text/javascript" src="${static.url('js/student_account/tos_modal.js')}"></script>
+    % endif
```
Just realized we will also need to add the changes to this template to the "themed" versions of this template in [edx/edx-themes](https://github.com/edx/edx-themes):

https://github.com/edx/edx-themes/blob/master/edx-platform/globalacademy.hms.harvard.edu/lms/templates/student_account/login_and_register.html

https://github.com/edx/edx-themes/blob/master/edx-platform/harvardxplus.harvard.edu/lms/templates/student_account/login_and_register.html

https://github.com/edx/edx-themes/blob/master/edx-platform/juilliardopenclassroom.com/lms/templates/student_account/login_and_register.html

https://github.com/edx/edx-themes/blob/master/edx-platform/mitprofessionalx.mit.edu/lms/templates/student_account/login_and_register.html

https://github.com/edx/edx-themes/blob/master/edx-platform/professionaleducation.wharton.upenn.edu/lms/templates/student_account/login_and_register.html

replaceafill replied:
@douglashall Ah ok. Will do.

replaceafill replied:
@douglashall Done. Please review https://github.com/edx/edx-themes/pull/511

douglashall commented:
@estute Would you mind helping us run the tests on this PR one last time?

douglashall commented:
@replaceafill Can you rebase/squash this one before I ask Test Engineering to rerun the tests on this one again?

@bradenmacdonald Is there some list we need to get replaceafill on so that the tests run automatically for him?

replaceafill commented:
@douglashall Rebase/squash done.

estute commented:
jenkins run all

itsjeyd commented:
@douglashall Looks like this is ready to flip the switch! :shipit:

bradenmacdonald commented:
@douglashall 

> @bradenmacdonald Is there some list we need to get replaceafill on so that the tests run automatically for him?

Yes there is - I will follow up on that.

gsong commented:
Jenkins ok to test
